File: library/src/main/java/com/getstream/sdk/chat/rest/controller/TokenAuthInterceptor.java
package com.getstream.sdk.chat.rest.controller;

import com.getstream.sdk.chat.BuildConfig;
import com.getstream.sdk.chat.StreamChat;
import com.getstream.sdk.chat.interfaces.CachedTokenProvider;
import com.getstream.sdk.chat.rest.response.ErrorResponse;
import com.getstream.sdk.chat.utils.Constant;
import com.getstream.sdk.chat.utils.Utils;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.concurrent.CountDownLatch;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class TokenAuthInterceptor implements Interceptor {

    private final String TAG = getClass().getSimpleName();

    private CachedTokenProvider tokenProvider;
    private String token;

    TokenAuthInterceptor(CachedTokenProvider tokenProvider) {
        this.tokenProvider = tokenProvider;
    }

    @NotNull
    @Override
    public Response intercept(Chain chain) throws IOException {
        final CountDownLatch latch = new CountDownLatch(token == null ? 1 : 0);

        if (token == null) {
            tokenProvider.getToken(token -> {
                this.token = token;
                latch.countDown();
            });
        }

        try {
            latch.await();
        } catch (InterruptedException e) {
            StreamChat.getLogger().logT(this, e);
        }

        Request request = chain.request();
        request = addTokenHeader(request);
        request = addVersionHeader(request);
        Response response = chain.proceed(request);

        // check the error and only hit this path if the token was expired (error response code)
        if (!response.isSuccessful()) {
            ErrorResponse err = ErrorResponse.parseError(response);
            if (err.getCode() == ErrorResponse.TOKEN_EXPIRED_CODE) {
                StreamChat.getLogger().logD(this, "Retrying new request");
                token = null; // invalidate local cache
                tokenProvider.tokenExpired();
                response.close();
                response = chain.proceed(request);
            }
        }

        return response;
    }

    private Request addTokenHeader(Request req) {
        return req.newBuilder().header("Authorization", token).build();
    }

    private Request addVersionHeader(Request req) {
        return req.newBuilder().header(Constant.HEADER_VERSION, Utils.version()).build();
    }

}

File: app/src/main/java/io/getstream/chat/android/core/poc/library/api/TokenAuthInterceptor.kt
package io.getstream.chat.android.core.poc.library.api

import android.util.Log
import io.getstream.chat.android.core.poc.library.CachedTokenProvider
import io.getstream.chat.android.core.poc.library.TokenProvider.TokenProviderListener
import io.getstream.chat.android.core.poc.library.gson.JsonParserImpl
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.CountDownLatch


class TokenAuthInterceptor internal constructor(
    private val tokenProvider: () -> CachedTokenProvider?,
    private val anonymousAuth: () -> Boolean,
    private val jsonParser: JsonParserImpl
) : Interceptor {

    private val TOKEN_EXPIRED_CODE = 40
    private val TAG = javaClass.simpleName
    private var token: String? = null

    override fun intercept(chain: Interceptor.Chain): Response {

        if (anonymousAuth()) {
            return chain.proceed(chain.request())
        } else {
            val latch = CountDownLatch(if (token == null) 1 else 0)
            if (token == null) {
                tokenProvider()?.getToken(object : TokenProviderListener {
                    override fun onSuccess(token: String) {
                        this@TokenAuthInterceptor.token = token
                        latch.countDown()
                    }

                })
            }
            try {
                latch.await()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            val request: Request = addTokenHeader(chain.request())
            var response: Response = chain.proceed(request)

            if (!response.isSuccessful) {

                val err = jsonParser.toError(response)
                if (err.streamCode == TOKEN_EXPIRED_CODE) {
                    Log.d(TAG, "Retrying new request")
                    token = null // invalidate local cache
                    tokenProvider()?.tokenExpired()
                    response.close()
                    response = chain.proceed(request)
                } else {
                    throw IOException(err.message, err.cause)
                }
            }
            return response
        }
    }

    private fun addTokenHeader(req: Request): Request {
        return req.newBuilder().header("Authorization", token!!).build()
    }

}
